devito/types/equation.py

```diff
@@ -97,7 +97,9 @@ def evaluate(self):
             # all rules to be expunged during this procress.
             rules = default_rules(eq, eq._symbolic_functions)
             try:
-                eq = eq.xreplace({**eq.substitutions.rules, **rules})
+                eq = eq.xreplace(rules)
+                custom_rules = eq.substitutions.update_rules(eq)
+                eq = eq.xreplace(custom_rules)
```

Comment:
I don't like two xreplace back to back. I'm confident there's a way of working around this overhead (i.e., just one replace)

Reply:
I don't think `equation.py` should need any changes at all. Rules can, and should, be structured such that `eq = eq.xreplace({**eq.substitutions.rules, **rules})` still deals with everything.